program/src/processor.rs

```diff
@@ -751,23 +802,121 @@ impl Processor {
+                    pool_stake_authority_info.key,
+                    vote_account_info.key,
+                ),
+                &[
+                    pool_stake_info.clone(),
+                    vote_account_info.clone(),
+                    clock_info.clone(),
+                    stake_history_info.clone(),
+                    stake_config_info.clone(),
+                    pool_stake_authority_info.clone(),
+                ],
+                stake_authority_signers,
+            )?;
+        }
+
+        // if pool is fully active, we can move stake to the main account and lamports to the on-ramp
+        if pool_stake_is_fully_active {
+            // determine excess lamports in the main account before we touch either of them
+            let pool_excess_lamports = pool_stake_info
+                .lamports()
+                .saturating_sub(pool_stake_state.delegation.stake)
+                .saturating_sub(pool_stake_meta.rent_exempt_reserve);
+
+            // if the on-ramp is fully active, move its stake to the main pool account
+            if is_stake_fully_active(&onramp_status) {
+                invoke_signed(
+                    &stake::instruction::move_stake(
+                        pool_onramp_info.key,
+                        pool_stake_info.key,
+                        pool_stake_authority_info.key,
+                        onramp_status.effective,
+                    ),
+                    &[
+                        pool_onramp_info.clone(),
+                        pool_stake_info.clone(),
+                        pool_stake_authority_info.clone(),
+                    ],
+                    stake_authority_signers,
+                )?;
+            }
+
+            // if there are any excess lamports to move to the on-ramp, move them
+            if pool_excess_lamports > 0 {
+                invoke_signed(
+                    &stake::instruction::move_lamports(
+                        pool_stake_info.key,
+                        pool_onramp_info.key,
+                        pool_stake_authority_info.key,
+                        pool_excess_lamports,
+                    ),
+                    &[
+                        pool_stake_info.clone(),
+                        pool_onramp_info.clone(),
+                        pool_stake_authority_info.clone(),
+                    ],
+                    stake_authority_signers,
+                )?;
+            }
+
+            // finally, delegate the on-ramp account if it has sufficient undelegated lamports
+            // if activating, this means more lamports than the current activating delegation
+            // in all cases, this means having enough to cover the minimum delegation
+            // we do nothing if partially active. by doing it here, we know it cannot be fully active
+            let onramp_non_rent_lamports = pool_onramp_info
+                .lamports()
+                .saturating_sub(onramp_rent_exempt_reserve);
```

Comment:
Sorry if this is dumb, because I always forget -- `lamports()` here *will* reflect the added and subtracted lamports, correct?

Reply:
`lamports()` is read-only hot view of the account, so it will reflect any previous mutations but not mutate the balance itself. if you mean the latter you might be thinking of the `checked_sub_lamports()` helper on `BorrowedAccount`